File: 2487-remove-nodes-from-linked-list/2487-remove-nodes-from-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def removeNodes(self, head):
        st = []
        while head:
            while st and st[-1].val < head.val:
                st.pop()
            st.append(head)
            head = head.next
        
        prev = None
        while st:
            curr = st.pop()
            curr.next = prev
            prev = curr
        
        return prev
        

File: 2487-remove-nodes-from-linked-list/2487-remove-nodes-from-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNodes(ListNode* head) {
        stack<ListNode*> st;
        while(head) {
            while(!st.empty() && st.top()->val < head->val) st.pop();
            st.push(head);
            head = head->next;
        }

        ListNode* prev = NULL;
        while(!st.empty()) {
            ListNode* curr = st.top();
            st.pop();
            curr->next = prev;
            prev = curr;
        }

        return prev;
    }
};